compiler/extensions/cpp/runtime/test/zserio/BitStreamReaderTest.cpp

```diff
@@ -26,30 +26,30 @@ class BitStreamReaderTest : public ::testing::Test
 
 const size_t BitStreamReaderTest::BUFFER_SIZE;
 
-TEST_F(BitStreamReaderTest, bitBufferCtor)
+TEST_F(BitStreamReaderTest, readUnalignedData)
 {
-    uint8_t data[] = {1, 2, 3};
+    // number expected to read at offset
+    const uint8_t TestValue = 123u;
 
-    zserio::BitBuffer buffer(data, 17u);
-    zserio::BitStreamReader reader(buffer);
+    for (int offset = 0; offset <= 8; ++offset) {
+        uint8_t data[] = {0x00, 0x00, 0x00};
 
-    ASSERT_EQ(17u, reader.getBufferBitSize());
-}
+        // write test value at offset to data buffer
+        data[offset / 8    ] |= TestValue >> (offset % 8);
+        data[offset / 8 + 1] |= TestValue << (8u - offset % 8);
 
-TEST_F(BitStreamReaderTest, readUnalignedData)
-{
-    /* 1bit = 0, 8bit = 8, 8bit = 1 */
-    uint8_t data[] = {0x04, 0x00, 0x80};
+        zserio::BitBuffer buffer(data, 8u + offset);
+        zserio::BitStreamReader reader(buffer);
 
-    zserio::BitBuffer buffer(data, 17u);
-    zserio::BitStreamReader reader(buffer);
```

Comment:
It's still slightly related to BitBuffer. When using:
```
zserio::BitStreamReader reader(data, (data + 8 + 7 / 8);
```
, then all the values seem to be read correctly. Only the EOF check fails. I'm just fixing similar issue in ASIL verison. @mikir can reuse this test once fixing the BitBuffer.

Reply:
I think the bug is, that zserio does not correctly handle the case of `ctx.bufferBitSize` not being a multiple of 8.

Reply:
Yes, you are correct :+1:. BitBuffer is quite new, and before introducing of the BitBuffer, the `ctx.bufferBitSize` was *always* rounded to full bytes. And the code in `loadCacheNext` is very very old. I have faced the same issue in ASIL repo few days ago, where we started to use the BitBuffer as an external buffer for writer and reader. I did very similar fix already, but since we were forced to split the repositories, it will take some time until it will bubble here ;-).

Please see my version of loadCacheNext:
```
    /** Loads next 32 bits to 32-bit bit-cache. */
    inline void loadCacheNext(ReaderContext& ctx, uint8_t numBits)
    {
        static const uint8_t cacheBitSize = sizeof(BaseType) * 8;
        BaseType& cacheBuffer = getCacheBuffer(ctx.cache);

        // buffer must be always available in full bytes, even if some last bits are not used
        const size_t fullBufferBitSize = (ctx.bufferBitSize + 7) & ~0x07;

        // ctx.bitIndex is always byte aligned and ctx.cacheNumBits is always zero in this call
        const size_t byteIndex = ctx.bitIndex >> 3;
        if (fullBufferBitSize >= ctx.bitIndex + cacheBitSize)
        {
            cacheBuffer =
#ifdef ZSERIO_RUNTIME_64BIT
                    parse64(ctx.buffer + byteIndex);
#else
                    parse32(ctx.buffer + byteIndex);
#endif
            ctx.cacheNumBits = cacheBitSize;
        }
        else
        {
            checkEof(ctx, numBits);

            ctx.cacheNumBits = static_cast<uint8_t>(fullBufferBitSize - ctx.bitIndex);

            // always aligned to full bytes and less than cacheBitSize
            switch (ctx.cacheNumBits)
            {
#ifdef ZSERIO_RUNTIME_64BIT
            case 56:
                cacheBuffer = parse56(ctx.buffer + byteIndex);
                break;
            case 48:
                cacheBuffer = parse48(ctx.buffer + byteIndex);
                break;
            case 40:
                cacheBuffer = parse40(ctx.buffer + byteIndex);
                break;
            case 32:
                cacheBuffer = parse32(ctx.buffer + byteIndex);
                break;
#endif
            case 24:
                cacheBuffer = parse24(ctx.buffer + byteIndex);
                break;
            case 16:
                cacheBuffer = parse16(ctx.buffer + byteIndex);
                break;
            default: // 8
                cacheBuffer = parse8(ctx.buffer + byteIndex);
                break;
            }
        }
    }
```

Reply:
And yet I'm still not sure that my fix it enough. It seems that it still misses one `checkEof` in the first branch of the `if`. Please note that all this stuff is a heavy performance optimization. And the more operations we will add, the less sense the optimization have.